### PR TITLE
migrating custom groups

### PR DESCRIPTION
**Pull request by ArtificialOwl**
### migrate owncloud's Custom Groups to Teams

for each Custom Group (CG):

- creating Team using CG oldest Admin as Owner
- adding CG members to generated Team
- leveling other Admins from CG as Admin in generated Team,
- converting shares to CG to generated Team,
- fixing children shares

### original data

```
+--------------+---------------------+------+-----+---------+----------------+
| Field        | Type                | Null | Key | Default | Extra          |
+--------------+---------------------+------+-----+---------+----------------+
| group_id     | bigint(20) unsigned | NO   | PRI | NULL    | auto_increment |
| uri          | varchar(255)        | NO   | UNI | NULL    |                |
| display_name | varchar(64)         | NO   |     | NULL    |                |
+--------------+---------------------+------+-----+---------+----------------+

+----------+---------------------+------+-----+---------+-------+
| Field    | Type                | Null | Key | Default | Extra |
+----------+---------------------+------+-----+---------+-------+
| group_id | bigint(20) unsigned | NO   | PRI | NULL    |       |
| user_id  | varchar(64)         | NO   | PRI | NULL    |       |
| role     | int(11)             | NO   |     | 0       |       |
+----------+---------------------+------+-----+---------+-------+
```

### CODE REVIEW

File: lib/Command/MigrateCustomGroups.php [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+			if ($ownerId === '' || $groupId === 0) {
+				continue; // if group or owner is not know, we ignore the entry.
+			}
+
+			// based on owner's userid, we create federateduser and a new circle
+			$this->output->writeln('+ New Team <info>' . $rowCG['display_name'] . '</info>, owner by <info>' . $ownerId . '</info>');
+			$owner = $this->cachedFed($ownerId);
+
+			$this->circlesManager->startSession($owner);
+			$circle = $this->circlesManager->createCircle($rowCG['display_name']);
```
Comment (mejo-) [inline lib/Command/MigrateCustomGroups.php:98]: This at least can throw `CircleNameTooShortException`, but probably it also throws if the name is already taken, no? We probably need some exception handling here.

**Review by mejo- — COMMENTED** [review]
I have no idea about owncloud and their custom groups functionality, but apart from one comment i didn't spot anything obvious :relaxed:

File: lib/Command/MigrateCustomGroups.php (hunk 1 above)
Comment (ArtificialOwl) [reply lib/Command/MigrateCustomGroups.php:84]: the too short exception might occurs, but in case of name taken a digit will be added

Comment (ArtificialOwl) [reply lib/Command/MigrateCustomGroups.php]: (added an exception management)
also details about original data in the top comment

**Review by mejo- — APPROVED** [review]
Code changes look sensible now. Thanks for adding additional context to the PR description. I didn't test this, bug I guess you did :wink:

**Pull request merged**